### PR TITLE
Return list of files preserving given pattern order on glob

### LINKED ISSUE #7518: Build failure triggered by reordered umbrella header file
* [X] I've read and understood the [*CONTRIBUTING guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?

` git clone git@github.com:paulb777/NVMDummyTestPod.git`
`~/Cocoapods/CocoaPods/bin/pod update --no-repo-update`
`open NVMDummy.xcworkspace/`
Build and notice build failure

## What did you expect to happen?

I expected the build to succeed like it does if I run `pod update` with the 1.4.0 version.  Running with [latest master](https://github.com/CocoaPods/CocoaPods/commit/606b943343b525418d4e24564a141973fe6cc263) fails.

The difference is the file `./Pods/Target Support Files/Protobuf/Protobuf-umbrella.h` which has the imports specified by a wildcard listed ahead of the explicitly named headers, despite the wildcard listed first in the podspec file.

### Stack

```
   CocoaPods : 1.4.0
        Ruby : ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin17]
    RubyGems : 2.6.13
        Host : Mac OS X 10.13.3 (17D47)
       Xcode : 9.2 (9C40b)
         Git : git version 2.15.0.403.gc27cc4dac6-goog
Ruby lib dir : /Users/paulbeusterien/.rbenv/versions/2.4.0/lib
Repositories : cpdc-eap-spec - sso://cpdc-eap/spec @ c8892452be7c52ff70c238c111369b92ab97e90c
               cpdc-internal-spec - sso://cpdc-internal/spec @ bba62265f71bd5cbed7df4a35a1c9abf8064f139
               master - https://github.com/CocoaPods/Specs.git @ afe932100bfb533402402b5f0d3fa9032bdca57b
               paul-specs - sso://user/paulbeusterien/Specs @ 0b01f6ce5731056ef13971c3c2e339fb1c789cff
```

### Installation Source

```
Executable Path: /Users/paulbeusterien/Cocoapods/CocoaPods/bin/pod
```

### Plugins

```
claide-plugins         : 0.9.2
cocoapods-deintegrate  : 1.0.2
cocoapods-dependencies : 1.0.0.beta.1
cocoapods-plugins      : 1.0.0
cocoapods-search       : 1.0.0
cocoapods-stats        : 1.0.0
cocoapods-trunk        : 1.3.0
cocoapods-try          : 1.1.0
cocoapods_debug        : 0.1.0
```

### Podfile

```ruby
use_frameworks!

source 'https://github.com/CocoaPods/Specs.git'


target 'NVMDummy' do
  platform :ios, '8.0'

  pod 'NVMDummy', :path => './'

end
```

## Project that demonstrates the issue

`git@github.com:paulb777/NVMDummyTestPod.git`

### PR DESCRIPTION
**Pull request by manuyavuz**
Closes #7518.

@paulb777 could you verify that this solves the issue?

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Return list of files preserving given pattern order on glob  
  [manuyavuz](https://github.com/manuyavuz)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by manuyavuz** [comment]
I guess CI failure is caused by not up-to-date integration specs.

**Comment by dnkoutso** [comment]
You do not need to bump the submodule, seems like you are pointing integration specs to an old sha.

File: lib/cocoapods/sandbox/path_list.rb [hunk 1]
```diff
@@ -129,23 +129,25 @@ def relative_glob(patterns, options = {})
...
-        end.flatten
-        list = exact_matches
-        unless all_patterns.empty?
-          list += full_list.select do |path|
-            all_patterns.any? do |p|
-              File.fnmatch(p, path, File::FNM_CASEFOLD | File::FNM_PATHNAME)
+        exact_matches = (full_list & patterns_array).to_set
+
+        unless patterns_array.empty?
+          list = patterns_array.map do |pattern|
```
Comment (segiddins) [inline lib/cocoapods/sandbox/path_list.rb]: `flat_map` instead of `flatten` ?

**Comment by dnkoutso** [comment]
@manuyavuz I pushed a commit to your branch that deals with the integration specs and also uses `flat_map` instead of `flatten`.

We aim to ship 1.5.0.beta.1 today so just wanted to get that in before we ship it!

**Comment by manuyavuz** [comment]
Oh I'm sorry, I just did the same before seeing your comment :(

**Comment by dnkoutso** [comment]
awesome!

**Pull request merged**